bug: Restructure PER-CS rule sets

Crell opened the pull request:
Resolves #6700.

This PR does a few things:

1. Sets the plan to have pinned versions of PER-CS as it evolves.  Right now there's just `PERCS10` (1.0), but as PER-CS 2.0 is now out, that can be added in a follow-up.
2. Introduces `PERCS` as a ruleset that just aliases to the latest release.  So when `PERCS20` is added, `PERCS` changes to point to that.
3. Makes `PER` an alias to `PERCS` for BC.  It can probably be deprecated and removed eventually, but I leave that to the maintainers to decide.

I don't know if there's any defaults elsewhere that should also be updated.  Again, I defer to the maintainers.

julienfalque commented on src/RuleSet/Sets/PERCS10Set.php:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @link https://github.com/php-fig/per-coding-style/blob/1.0.0/spec.md
+ */
+final class PERCS10Set extends AbstractRuleSetDescription
```
I think the name is confusing: is it PER CS 1.0 ou 10? I would override `getName()` to remove the ambiguity.

julienfalque commented on src/RuleSet/Sets/PERCS10Set.php:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @internal
+ *
+ * PER Coding Style v1.0.0.
```
Would a patch version of the recommendation require a new ruleset?

Crell replied:
Presumably not.  Dropping the .z here makes sense, I think.

julienfalque commented on src/RuleSet/Sets/PERSet.php:
```diff
@@ -26,12 +26,12 @@ final class PERSet extends AbstractRuleSetDescription
     public function getDescription(): string
     {
-        return 'Rules that follow `PER Coding Style <https://www.php-fig.org/per/coding-style/>`_.';
+        return 'Alias for the PERCS rules.  It is recommended you use @PERCS instead.';
```
Extra space:
```suggestion
        return 'Alias for the PERCS rules. It is recommended you use @PERCS instead.';
```

Crell replied:
2spaces4eva.

julienfalque commented on src/RuleSet/Sets/PERCS10Set.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function getDescription(): string
+    {
+        return 'Rules that follow `PER Coding Style 1.0 <https://www.php-fig.org/per/coding-style/>`_.';
```
Not sure this method is always called in a reStructuredText context. But since this is quite legible, maybe that's ok anyway.

Crell replied:
No idea.  That's how the existing `PERSet` was written, so I just copied-and-modified it.  Happy to change it to whatever.

julienfalque replied:
Right, let's keep it.

Crell commented:
I've updated the names.  Let me know if I did it right or if there's some other format I should use there.  I wasn't clear if spaces were allowed.

I also realized I missed the "risky" sets before, so I added those as well.

Crell commented:
Hm.  Did I not use the names correctly?  Please advise.

kubawerlos commented on src/RuleSet/Sets/PERCSSet.php:
```diff
@@ -0,0 +1,46 @@
+{
+    public function getName(): string
+    {
+        return 'PER-CS';
```
```suggestion
        return '@PER-CS';
```

> Hm. Did I not use the names correctly? Please advise.

@Crell  like this?

Crell replied:
Let's see if the bot likes it.

Crell commented:
OK, I don't understand what the tests are objecting to now.  Can someone help me out here?  I'm confused as to what I'm still screwing up. :smile:

kubawerlos commented:
Isn't the name of the test - `testFixerDocumentationFileIsUpToDate` - helping? Have you read https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/CONTRIBUTING.md? My bet would be you haven't run `php dev-tools/doc.php`.

BTW it is not illegal to run tests locally 😉

keradus commented on src/RuleSet/Sets/PERCS10RiskySet.php:
```diff
@@ -0,0 +1,44 @@
+ */
+final class PERCS10RiskySet extends AbstractRuleSetDescription
+{
+    public function getName(): string
```
I get the idea behind overriding this method, but I think it shows the problem we are bypassing here.
looking at class name, we still do now know if it's PERCS 1.0 or 10. I wonder if we can adjust class name too?

Crell replied:
I am open to suggestions, given that we can't have periods in class names.  Using getName() is what @julienfalque suggested.  I have no particular preference, so whatever works.

keradus replied:
`PERCS1x0RiskySet`? and `AbstractRuleSetDescription` can replace `\dx\d` with `\1.\2`

Wirone replied:
Let's use hardcoded values - we can easily change it in the future without breaking change, because name won't change, even if it's calculated automagically from class name.

PS. I removed that `Preg::replace()` which was added before at some point, because it wasn't used. If we actually need it, we can add it in separate PR.

keradus commented on src/RuleSet/Sets/PERCSRiskySet.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @link https://www.php-fig.org/per/coding-style/
+ *
+ * This Ruleset always points to the latest version of PER-CS.
```
I wonder if we can have test for that?
I imagine someone creating PERCS1.2, but forgetting to update definition of PERCS

Crell replied:
Off the cuff guess: A test that scans the folder for class names, filters for `PERCS`, orders alphabetically, and pops the first one?  Then use that to see if it resolves to the same name that `PERCSSet` points to?  Just spitballing.

keradus replied:
i would only replace first part, you can use https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/RuleSet/RuleSets.php#L55 instead of scanning the folder

keradus commented on src/RuleSet/Sets/PERSet.php:
```diff
@@ -19,19 +19,19 @@
 /**
  * @internal
```
```suggestion
 * @internal
 * @deprecated
 * @TODO 4.0 remove me
```

Crell commented:
OK, I *think* I got it this time.  Let's see what the bot thinks.

Crell commented:
Anyone else still interested in this PR, now that it's actually passing? :smile:

keradus commented:
you put `per-cs1.0` in test fixture name, but keep `PERCS10Set` for Set classname file.
what do you think about using `1x0` in filename to indicate the dot? ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6707#discussion_r1060215806

kubawerlos commented on src/RuleSet/Sets/PERCS1x0Set.php:
```diff
@@ -0,0 +1,44 @@
+
+    public function getDescription(): string
+    {
+        return 'Rules that follow `PER Coding Style 1.0 <https://www.php-fig.org/per/coding-style/>`_.';
```
Is the `_` expected to be here?

Crell replied:
It was in the PERSet class already, so I just kept it: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/RuleSet/Sets/PERSet.php#L35

Whether that was correct in the first place, I cannot say.  I just followed the existing pattern.

kubawerlos replied:
Sorry for the fuss, apparently that how links work in RST files :)

Crell commented:
Anything else that needs to be done here?

viktorprogger commented:
It's released today! But it's 2.0 instead of 1.1.
https://github.com/php-fig/per-coding-style/releases

Crell commented:
All the more reason we should get this code merged soon so work can start on the 2x0 ruleset... 😄 

I'm not going to try and fix the conflicts until I hear from someone with merge access that this approach is acceptable.  No sense doing (more) work if it's not going to be accepted.

Wirone commented:
@Crell I really would like to get it merged, but I need some time to get familiar with it. I believe all @keradus' doubts were addressed, @kubawerlos gave approval, for me it looks good at first glance (deprecate too general ruleset + introduce version specific rulesets), I just need to dig deeper into provided changes and previous discussions.

If you don't mind, I can rebase it to get rid of the conflicts and to trigger new CI pipeline. But of course you can do it too 🙂. Let me know, thanks in advance.

Crell commented:
You'll probably have a chance to rebase it before I will, so feel free.  I am happy to answer questions about it, though "why is it done that way?" is mostly answered by "because some other PR reviewer said to and I don't know the tool well enough to disagree." 😄 

(Also updated the description to account for there now being a PER-CS 2, not 1.1.)

Wirone commented on src/RuleSet/Sets/PERCSRiskySet.php:
```diff
@@ -0,0 +1,46 @@
+
+    public function getDescription(): string
+    {
+        return 'Rules that follow the latest `PER Coding Style <https://www.php-fig.org/per/coding-style/>`_.';
```
```suggestion
        return 'Risky rules that follow the latest `PER Coding Style <https://www.php-fig.org/per/coding-style/>`_.';
```

Wirone commented on src/RuleSet/Sets/PERCSSet.php:
```diff
@@ -0,0 +1,46 @@
+ * @see https://www.php-fig.org/per/coding-style/
+ *
+ * This Ruleset always points to the latest version of PER-CS.
+ * To fix on a specific version of PER-CS, use `PERCS10` (the 1.0 release),
+ * `PERCS11` (the 1.1 release), etc.
```
```suggestion
 * To fix on a specific version of PER-CS, use `PER-CS1.0` (the 1.0 release) etc.
```

Wirone commented on src/RuleSet/Sets/PERCSRiskySet.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @see https://www.php-fig.org/per/coding-style/
+ *
+ * This Ruleset always points to the latest version of PER-CS.
+ * To fix on a specific version of PER-CS, use `PERCS10` (the 1.0 release),
+ * `PERCS11` (the 1.1 release), etc.
```
```suggestion
 * This Ruleset always points to the latest version of PER-CS risky rules.
 * To fix on a specific version of PER-CS risky rules, use `PER-CS1.0:risky` etc.
```

Wirone commented on src/RuleSet/Sets/PERCSSet.php:
I am not sure if it's a good idea to provide dynamic alias for latest PER-CS. It has several drawbacks:

- we need to [keep track of built-in sets](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6707#discussion_r1058584504) to prevent aliases getting out of sync
- we need to duplicate integration tests for the alias and will need to change them every time when new version of PER-CS is released
- if people out there use `@PER-CS` in their config it can suddenly change its behavior after Fixer upgrade. I believe people would rather prefer to change their config knowingly. It would be better to warn people somehow that they use outdated PER-CS.

Crell replied:
I'm comfortable either way.  I can see the argument for it and against, and I'd probably use the pinned version anyway myself, so I defer to the maintainers.

kubawerlos replied:
- I can imagine someone out of laziness/easiness would like to have `@PER-CS-lastest` (or whatever the name would be) and be rather prompt to re-format the code over remembering to update the set name every time a new version f PER-CS is released.
- we can skip the integration tests for the alias of the latest and instead have a test that it is indeed an alias of the latest version

➕ 1️⃣  to @Crell "I'd probably use the pinned version anyway myself", so maybe we can add an alias in the future if there is a big need for it.

Wirone reviewed (changes requested):
General idea is OK for me, but I have some doubts, 2 most important are:

- [how to handle `PER-CS*` naming](#discussion_r1199502500)
- [should we provide alias for latest `PER-CS`?](#discussion_r1199507132)

Crell commented on src/RuleSet/Sets/PERCS1x0RiskySet.php:
```diff
@@ -0,0 +1,44 @@
+use PhpCsFixer\RuleSet\AbstractRuleSetDescription;
+
+/**
+ * @internal
```
I think the intent on that one was to again discourage people from using `PER` specifically, since that should be deprecated and removed.  I may have over-interpreted it to include elsewhere.  We can remove them.

kubawerlos replied:
The classes are internal in terms of PHP API, so we can freely change it's signature.

Wirone replied:
But it does not make sense for me.. internal, but for public usage? 🤷‍♂️

kubawerlos replied:
For public usage via configuration as `@Symfony` set etc., the end user does not care how is the implementation made.

Wirone replied:
It makes sense 👌. Let's keep it.

Crell commented on src/RuleSet/AbstractRuleSetDescription.php:
```diff
@@ -26,6 +28,7 @@ public function __construct()
     public function getName(): string
     {
         $name = substr(static::class, 1 + strrpos(static::class, '\\'), -3);
+        $name = Preg::replace('/\dx\d/', '\1.\2', $name);
```
I really don't care about the class naming so I'm happy with basically anything, as long as it's clear which PER-CS version it references.

Wirone commented:
@Crell are you able to address last 2 concerns ([this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6707#discussion_r1200855701) and [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6707#discussion_r1200862169))? We could finish it 🙂.

Crell commented:
I thought you had taken over. 😄   Here's a stab at it.  Let's see if the bot approves.

Once this is done, we should figure out how to get PER-CS 2 implemented.  There are probably going to need to be a few new rules.

Wirone commented on src/RuleSet/AbstractRuleSetDescription.php:
```diff
@@ -28,7 +28,8 @@ public function __construct()
         $name = substr(static::class, 1 + strrpos(static::class, '\\'), -3);
-        $name = Preg::replace('/\dx\d/', '\1.\2', $name);
+        $name = Preg::replace('/(\d)x(\d)/', '\1.\2', $name);
+        $name = str_replace('_', '-', $name);
```
All changes in `AbstractRuleSetDescription` were reverted.

Wirone commented:
I am not 100% sure about current file/class naming (`PERCS1x0Set` and `PERCS1x0RiskySet`), because looking at other sets none of them uses `NxN` notation - both PHP and PHPUnit related sets have only numbers (like `PHPUnit100MigrationRiskySet`, but I believe we can change those files/classes names when needed, leaving actual set's name as-is. It is something that should be standardise somehow, but it's out of scope of this PR.

So.. I'm going to merge it and if anything looks wrong to @keradus we can change it later, before release.

Wirone commented:
Thank you very much @Crell 🍻

Crell commented:
Excellent!  Now someone can start on the PER-CS 2.0 ruleset. 😄
